File: Data Structures/1. CPP/a. Arrays.cpp
#include <iostream>
#include <vector>

template <typename T>
class CustomArray {
private:
    std::vector<T> elements;  // Internal storage for elements

public:
    // Initialize the custom array with a specified initial size
    CustomArray(int initialSize) {
        elements.resize(initialSize);
    }

    // Append an element to the end of the custom array
    void append(T element) {
        elements.push_back(element);
    }

    // Insert an element at the specified index
    void insert(T element, int index) {
        elements.insert(elements.begin() + index, element);
    }

    // Remove the first occurrence of the specified element
    void remove(T element) {
        auto it = std::find(elements.begin(), elements.end(), element);
        if (it != elements.end()) {
            elements.erase(it);
        }
    }

    // Remove and return the element at the specified index
    T pop(int index) {
        if (index >= 0 && index < elements.size()) {
            T element = elements[index];
            elements.erase(elements.begin() + index);
            return element;
        }
        return nullptr; // Returning nullptr in C++ isn't valid, use appropriate handling for the type T
    }

    // Get the element at the specified index
    T get(int index) {
        if (index >= 0 && index < elements.size()) {
            return elements[index];
        }
        return nullptr; // Returning nullptr in C++ isn't valid, use appropriate handling for the type T
    }

    // Set the element at the specified index to the given element
    void set(T element, int index) {
        if (index >= 0 && index < elements.size()) {
            elements[index] = element;
        }
    }

    // Return the current size of the custom array
    int size() {
        return elements.size();
    }

    // Check if the custom array is empty
    bool isEmpty() {
        return elements.empty();
    }

    // Return the index of the first occurrence of the specified element, or -1 if not found
    int index(T element) {
        auto it = std::find(elements.begin(), elements.end(), element);
        if (it != elements.end()) {
            return std::distance(elements.begin(), it);
        }
        return -1;
    }

    // Return the number of occurrences of the specified element in the custom array
    int count(T element) {
        return std::count(elements.begin(), elements.end(), element);
    }

    // Reverse the order of elements in the custom array
    void reverse() {
        std::reverse(elements.begin(), elements.end());
    }

    // Sort the elements in the custom array in ascending order
    void sort() {
        std::sort(elements.begin(), elements.end());
    }

    // Return a new array containing elements from the specified start index to the end index
    std::vector<T> slice(int start, int end) {
        if (start >= 0 && end < elements.size() && start <= end) {
            return std::vector<T>(elements.begin() + start, elements.begin() + end + 1);
        }
        return std::vector<T>();
    }

    // Append elements from another custom array to the end of this custom array
    void extend(const CustomArray& otherArray) {
        elements.insert(elements.end(), otherArray.elements.begin(), otherArray.elements.end());
    }

    // Remove all elements from the custom array
    void clear() {
        elements.clear();
    }

    // Check if the custom array contains the specified element
    bool contains(T element) {
        return std::find(elements.begin(), elements.end(), element) != elements.end();
    }

    // Return a new array with only unique elements
    std::vector<T> unique() {
        std::vector<T> uniqueElements(elements.begin(), elements.end());
        std::sort(uniqueElements.begin(), uniqueElements.end());
        auto last = std::unique(uniqueElements.begin(), uniqueElements.end());
        uniqueElements.erase(last, uniqueElements.end());
        return uniqueElements;
    }

    // Apply a transformation function to each element of the custom array
    template <typename U>
    std::vector<U> map(U (*transform)(T)) {
        std::vector<U> result;
        for (const auto& element : elements) {
            result.push_back(transform(element));
        }
        return result;
    }

    // Return a new array with elements that satisfy the given filtering function
    std::vector<T> filter(bool (*isIncluded)(T)) {
        std::vector<T> result;
        for (const auto& element : elements) {
            if (isIncluded(element)) {
                result.push_back(element);
            }
        }
        return result;
    }

    // Parameter closure: A function that takes an element of the array as a parameter.
    void forEach(void (*closure)(T)) {
        for (const auto& element : elements) {
            closure(element);
        }
    }
};

int main() {
    CustomArray<int> customArray(5);
    customArray.append(10);
    customArray.append(20);
    customArray.insert(15, 1);

    std::cout << "Size of array: " << customArray.size() << "\n";

    // Print each element
    customArray.forEach([](int element) {
        std::cout << element << " ";
    });

    return 0;
}

File: data_structures/4_swift/a_arrays.swift
struct CustomArray<T> {
    private var elements: [T]  // Internal storage for elements

    // Initialize the custom array with a specified initial size
    init(initialSize: Int) {
        elements = Array(repeating: nil, count: initialSize)
    }

    // Append an element to the end of the custom array
    mutating func append(_ element: T) {
        elements.append(element)
    }

    // Insert an element at the specified index
    mutating func insert(_ element: T, at index: Int) {
        elements.insert(element, at: index)
    }

    // Remove the first occurrence of the specified element
    mutating func remove(_ element: T) {
        if let index = elements.firstIndex(of: element) {
            elements.remove(at: index)
        }
    }

    // Remove and return the element at the specified index
    mutating func pop(at index: Int) -> T? {
        guard index >= 0, index < elements.count else { return nil }
        return elements.remove(at: index)
    }

    // Get the element at the specified index
    func get(at index: Int) -> T? {
        guard index >= 0, index < elements.count else { return nil }
        return elements[index]
    }

    // Set the element at the specified index to the given element
    mutating func set(_ element: T, at index: Int) {
        guard index >= 0, index < elements.count else { return }
        elements[index] = element
    }

    // Return the current size of the custom array
    func size() -> Int {
        return elements.count
    }

    // Check if the custom array is empty
    func isEmpty() -> Bool {
        return elements.isEmpty
    }

    // Return the index of the first occurrence of the specified element, or nil if not found
    func index(of element: T) -> Int? {
        return elements.firstIndex(of: element)
    }

    // Return the number of occurrences of the specified element in the custom array
    func count(of element: T) -> Int {
        return elements.filter { $0 == element }.count
    }

    // Reverse the order of elements in the custom array
    mutating func reverse() {
        elements.reverse()
    }

    // Sort the elements in the custom array in ascending order
    mutating func sort() {
        elements.sort(by: { (first, second) -> Bool in
            if let firstComparable = first as? Comparable, let secondComparable = second as? Comparable {
                return firstComparable < secondComparable
            }
            return false
        })
    }

    // Return a new array containing elements from the specified start index to the end index
    func slice(start: Int, end: Int) -> [T] {
        guard start >= 0, end < elements.count, start <= end else { return [] }
        return Array(elements[start...end])
    }

    // Append elements from another custom array to the end of this custom array
    mutating func extend(_ otherArray: CustomArray) {
        elements.append(contentsOf: otherArray.elements)
    }

    // Remove all elements from the custom array
    mutating func clear() {
        elements.removeAll()
    }

    // Check if the custom array contains the specified element
    func contains(_ element: T) -> Bool {
        return elements.contains(element)
    }

    // Return a new array with only unique elements
    func unique() -> [T] {
        return Array(Set(elements))
    }

    // Apply a transformation function to each element of the custom array
    mutating func map<U>(_ transform: (T) -> U) -> [U] {
        return elements.map(transform)
    }

    // Return a new array with elements that satisfy the given filtering function
    func filter(_ isIncluded: (T) -> Bool) -> [T] {
        return elements.filter(isIncluded)
    }

    // Parameter closure: A closure that takes an element of the array as a parameter.
    func forEach(_ closure: (T) -> Void) {
        for element in elements {
            closure(element)
        }
    }
}
